test/support/generators.ex

```diff
@@ -0,0 +1,293 @@
+       * e.g. `["null", "integer", "boolean"]`
+  """
+  def schematic_and_data(opts \\ []) do
+    StreamData.bind(data(0, opts), fn raw_data ->
+      StreamData.constant({schematic_from_data(raw_data) |> Enum.fetch!(1), raw_data})
+    end)
+  end
+
+  defp data(depth, opts \\ [])
+
+  defp data(8, opts) do
+    using = Keyword.get(opts, :using, [])
+    excluding = Keyword.get(opts, :excluding, [])
+
+    filter =
+      case using do
+        [] -> {:reject, excluding}
+        _ -> {:accept, using}
+      end
+
+    ["integer", "boolean", "null", "string", "list", "tuple", "map"]
+    |> then(fn kinds ->
+      case filter do
+        {:reject, excluding} -> Enum.reject(kinds, &Enum.member?(excluding, &1))
+        {:accept, using} -> Enum.filter(kinds, &Enum.member?(using, &1))
+      end
+    end)
+    |> Enum.map(fn
+      "integer" ->
+        StreamData.integer()
+
+      "boolean" ->
+        StreamData.boolean()
+
+      "null" ->
+        StreamData.constant(nil)
+
+      "string" ->
+        StreamData.binary()
+
+      "tuple" ->
+        StreamData.list_of(scalar(), max_length: 5) |> StreamData.map(&List.to_tuple/1)
+
+      "list" ->
+        StreamData.list_of(scalar(), max_length: 5)
+
+      "map" ->
+        StreamData.map_of(
+          StreamData.one_of([StreamData.binary(), StreamData.integer()]),
+          scalar(),
+          max_length: 5
+        )
+    end)
+    |> StreamData.one_of()
+  end
+
+  defp data(depth, opts) do
+    using = Keyword.get(opts, :using, [])
+    excluding = Keyword.get(opts, :excluding, [])
+
+    filter =
+      case using do
+        [] -> {:reject, excluding}
+        _ -> {:accept, using}
+      end
+
+    ["integer", "boolean", "null", "string", "list", "tuple", "map"]
+    |> then(fn kinds ->
+      case filter do
+        {:reject, excluding} -> Enum.reject(kinds, &Enum.member?(excluding, &1))
+        {:accept, using} -> Enum.filter(kinds, &Enum.member?(using, &1))
+      end
+    end)
+    |> Enum.map(fn
+      "integer" ->
+        StreamData.integer()
+
+      "boolean" ->
+        StreamData.boolean()
+
+      "null" ->
+        StreamData.constant(nil)
+
+      "string" ->
+        StreamData.binary()
+
+      "tuple" ->
+        StreamData.list_of(data(depth + 1), max_length: 5) |> StreamData.map(&List.to_tuple/1)
+
+      "list" ->
+        StreamData.list_of(data(depth + 1), max_length: 5)
+
+      "map" ->
+        StreamData.map_of(
+          StreamData.one_of([StreamData.binary(), StreamData.integer()]),
+          data(depth + 1),
+          max_length: 5
+        )
+    end)
+    |> StreamData.one_of()
+  end
+
+  defp schematic_from_data(data) when is_integer(data),
+    do:
```

Comment:
sort of a mitch-ism, but i heavily disklike these multi-line keyword style function clauses.

if the body is more than one line, always do `do`/`end`

Reply:
makes sense. each of these started as one-liners and then grew so i don't think i noticed that the shape of the function looked kinda wack